### PR TITLE
Add tty as separation option in container/service exec

### PR DESCRIPTION
**Pull request by jakolehm**
This has one breaking change, now `-i` option means `--interactive` in `service exec` (to match docker etc tools behaviour).

### CODE REVIEW

**Comment by jnummelin** [comment]
Something flaky in piping input from another command:
```
$ echo 'saippuakauppias' | k service exec -i nginx rev
saippuakauppias

$ echo 'saippuakauppias' | k service exec -i nginx rev

```

Notice the second command output being empty.

On agent logs, following is seen:
```
kontena-agent | D, [2017-06-04T20:22:51.040040 #1] DEBUG -- Kontena::RpcServer: rpc request: Kontena::Rpc::DockerContainerApi#create_exec ["682b4cfa5e08fa383a166fe533629f4711c81a74b106070b1254d8c98d9e510e"]
kontena-agent | I, [2017-06-04T20:22:51.043486 #1]  INFO -- Kontena::Actors::ContainerExec: initialized (session 665f0424-3f80-47f3-bbb1-e69046ea9b1d)
kontena-agent | D, [2017-06-04T20:22:51.066622 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#run_exec ["665f0424-3f80-47f3-bbb1-e69046ea9b1d", ["rev"], false, true]
kontena-agent | I, [2017-06-04T20:22:51.067922 #1]  INFO -- Kontena::Actors::ContainerExec: starting command: ["rev"] (tty: false, stdin: true)
kontena-agent | D, [2017-06-04T20:22:51.082201 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["665f0424-3f80-47f3-bbb1-e69046ea9b1d", nil]
kontena-agent | D, [2017-06-04T20:22:51.082623 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["665f0424-3f80-47f3-bbb1-e69046ea9b1d", "saippuakauppias\n"]
kontena-agent | E, [2017-06-04T20:22:51.083488 #1] ERROR -- : Actor crashed!
kontena-agent | IOError: closed stream
kontena-agent | 	/app/lib/kontena/actors/container_exec.rb:37:in `write'
kontena-agent | 	/app/lib/kontena/actors/container_exec.rb:37:in `input'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:74:in `block (2 levels) in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:97:in `exclusive'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid.rb:421:in `exclusive'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:74:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
kontena-agent | I, [2017-06-04T20:22:51.083601 #1]  INFO -- Kontena::Actors::ContainerExec: command finished: ["rev"] with code 0
```

Comparing agent logs to successful exec:
```
kontena-agent | D, [2017-06-04T20:22:49.327298 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["87fd34c5-a66c-4c6a-aaa8-e9c13d252916", "saippuakauppias\n"]
kontena-agent | D, [2017-06-04T20:22:49.327850 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["87fd34c5-a66c-4c6a-aaa8-e9c13d252916", nil]
```

It seems really 50/60 whether agent gets the `nil` before the actual piped input or not. And when `nil` hits agent first, it closes the input stream(?) and subsequent input crashes the actor.

**Comment by jnummelin** [comment]
Could it be that the WS "piping" through server mangles the input ordering:
```
$ for i in {1..10}; do echo $i; done | k service exec -i --shell nginx echo
```
Produces in agent:
```
kontena-agent | D, [2017-06-04T21:24:33.229369 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#run_exec ["030f54e2-524d-4a51-9c84-f062fbf7e89f", ["/bin/sh", "-c", "echo"], false, true]
kontena-agent | I, [2017-06-04T21:24:33.229921 #1]  INFO -- Kontena::Actors::ContainerExec: starting command: ["/bin/sh", "-c", "echo"] (tty: false, stdin: true)
kontena-agent | D, [2017-06-04T21:24:33.238428 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "3\n"]
kontena-agent | D, [2017-06-04T21:24:33.239035 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "1\n"]
kontena-agent | D, [2017-06-04T21:24:33.239433 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "4\n"]
kontena-agent | D, [2017-06-04T21:24:33.240158 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "2\n"]
kontena-agent | D, [2017-06-04T21:24:33.242675 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "8\n"]
kontena-agent | D, [2017-06-04T21:24:33.243268 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "6\n"]
kontena-agent | D, [2017-06-04T21:24:33.243603 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "7\n"]
kontena-agent | D, [2017-06-04T21:24:33.244398 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "9\n"]
kontena-agent | D, [2017-06-04T21:24:33.245881 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "10\n"]
kontena-agent | D, [2017-06-04T21:24:33.246195 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", nil]
kontena-agent | D, [2017-06-04T21:24:33.246848 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::DockerContainerApi#tty_input ["030f54e2-524d-4a51-9c84-f062fbf7e89f", "5\n"]
kontena-agent | E, [2017-06-04T21:24:33.249795 #1] ERROR -- : Actor crashed!
kontena-agent | IOError: closed stream
kontena-agent | 	/app/lib/kontena/actors/container_exec.rb:37:in `write'
kontena-agent | 	/app/lib/kontena/actors/container_exec.rb:37:in `input'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:74:in `block (2 levels) in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:97:in `exclusive'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid.rb:421:in `exclusive'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:74:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
kontena-agent | I, [2017-06-04T21:24:33.250216 #1]  INFO -- Kontena::Actors::ContainerExec: command finished: ["/bin/sh", "-c", "echo"] with code 0
```
Note the mangled sequence on input. :/

File: test/spec/features/service/exec_spec.rb [hunk 1]
```diff
@@ -43,6 +43,12 @@
     expect(k.run).to be_truthy
   end
 
+  it 'runs a command with piped stdin' do
+    k = kommando("$ echo beer | kontena service exec -i test-1 rev")
```
Comment (jnummelin) [inline test/spec/features/service/exec_spec.rb]: Why is there `$` in the test command? It seems to work though :)

File: agent/lib/kontena/actors/container_exec.rb [hunk 2]
```diff
@@ -15,43 +15,59 @@ def initialize(container)
...
       info "initialized (session #{@uuid})"
+
+      every(60) { self.check_activity }
     end
 
-    # @param [String] input
-    def input(input)
-      @write_pipe.write(input)
+    def check_activity
+      if @last_input < (Time.now.to_i - 60*60)
```
Comment (jnummelin) [inline agent/lib/kontena/actors/container_exec.rb]: The possible ping-pong can be implemented in a later PR as an enhancement.

File: test/spec/features/service/exec_spec.rb (hunk 1 above)
Comment (jakolehm) [inline test/spec/features/service/exec_spec.rb]: It runs the command in a shell (kommando feature).

Comment (jnummelin) [inline test/spec/features/service/exec_spec.rb]: ah, nice. tried to look in kommando quickly but didn't find it :)

**Pull request merged**